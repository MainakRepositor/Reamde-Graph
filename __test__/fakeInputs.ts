import { graphArgs } from '../interfaces/interface';

export let fakeQueryString = [
  {
    username: 'githubusername',
  },
  {
    username: 'githubusername',
    bg_color: '44475a',
  },
  {
    username: 'githubusername',
    bg_color: '44475a',
    color: '000000',
  },
  {
    username: 'githubusername',
    bg_color: '44475a',
    color: '000000',
    line: '9e4c98',
  },
  {
    username: 'githubusername',
    bg_color: '44475a',
    color: '000000',
    line: '9e4c98',
    point: '44475a',
  },
  {
    username: 'githubusername',
    bg_color: '44475a',
    color: 'f8f8f2',
    line: 'ff79c6',
    point: 'bd93f9',
    area: true,
  },
  {
    username: 'githubusername',
    bg_color: '44475a',
    color: 'f8f8f2',
    line: 'ff79c6',
    point: 'bd93f9',
    area: true,
    hide_border: true,
  },
];

export let fakeQueryStringRes = [
  {
    username: 'githubusername',
    colors: {
      bgColor: 'ffcfe9',
      borderColor: 'ffffff',
      color: '9e4c98',
      lineColor: '9e4c98',
      pointColor: '403d3d',
    },
    area: false,
  },
  {
    username: 'githubusername',
    colors: {
      bgColor: '44475a',
      borderColor: 'ffffff',
      color: '9e4c98',
      lineColor: '9e4c98',
      pointColor: '403d3d',
    },
    area: false,
  },
  {
    username: 'githubusername',
    colors: {
      bgColor: '44475a',
      borderColor: 'ffffff',
      color: '000000',
      lineColor: '9e4c98',
      pointColor: '403d3d',
    },
    area: false,
  },
  {
    username: 'githubusername',
    colors: {
      bgColor: '44475a',
      borderColor: 'ffffff',
      color: '000000',
      lineColor: '9e4c98',
      pointColor: '403d3d',
    },
    area: false,
  },
  {
    username: 'githubusername',
    colors: {
      bgColor: '44475a',
      borderColor: 'ffffff',
      color: '000000',
      lineColor: '9e4c98',
      pointColor: '44475a',
    },
    area: false,
  },
  {
    username: 'githubusername',
    colors: {
      bgColor: '44475a',
      borderColor: 'ffffff',
      color: 'f8f8f2',
      lineColor: 'ff79c6',
      pointColor: 'bd93f9',
    },
    area: true,
  },
  {
    username: 'githubusername',
    colors: {
      bgColor: '44475a',
      borderColor: '0000',
      color: 'f8f8f2',
      lineColor: 'ff79c6',
      pointColor: 'bd93f9',
    },
    area: true,
  },
];

export let fakeGraphArgs: graphArgs = {
  height: 10,
  width: 20,
  colors: {
    bgColor: '44475a',
    borderColor: '0000',
    color: 'f8f8f2',
    lineColor: 'ff79c6',
    pointColor: 'bd93f9',
  },
  title: 'graphSvg',
  line: Promise.resolve('line'),
};

export let options = {
  width: 10,
  height: 5,
  axisY: {
    title: 'Contributions',
    onlyInteger: true,
    offset: 70,
    labelOffset: {
      y: 4.5,
    },
  },
  axisX: {
    title: 'Days',
    offset: 50,
    labelOffset: {
      x: -4.5,
    },
  },
  chartPadding: {
    top: 80,
    right: 50,
    bottom: 20,
    left: 20,
  },
  showArea: false,
  fullWidth: true,
};

export const expectedQuery = (username: string) => {
  return {
    query: `
      query userInfo($LOGIN: String!) {
       user(login: $LOGIN) {
         name
         contributionsCollection {
           contributionCalendar {
              totalContributions
              weeks {
                contributionDays {
                  contributionCount
                }
              }
            }
          }
        }
      },
    `,
    variables: {
      LOGIN: username,
    },
  };
};

export const dummyWeeksData = [
  {
    contributionDays: [
      {
        contributionCount: 2,
      },
      {
        contributionCount: 3,
      },
      {
        contributionCount: 16,
      },
      {
        contributionCount: 0,
      },
      {
        contributionCount: 5,
      },
      {
        contributionCount: 3,
      },
      {
        contributionCount: 0,
      },
    ],
  },
  {
    contributionDays: [
      {
        contributionCount: 0,
      },
      {
        contributionCount: 0,
      },
      {
        contributionCount: 0,
      },
      {
        contributionCount: 0,
      },
      {
        contributionCount: 2,
      },
      {
        contributionCount: 0,
      },
      {
        contributionCount: 5,
      },
    ],
  },
  {
    contributionDays: [
      {
        contributionCount: 1,
      },
      {
        contributionCount: 8,
      },
      {
        contributionCount: 5,
      },
      {
        contributionCount: 0,
      },
      {
        contributionCount: 0,
      },
      {
        contributionCount: 0,
      },
      {
        contributionCount: 0,
      },
    ],
  },
  {
    contributionDays: [
      {
        contributionCount: 0,
      },
      {
        contributionCount: 4,
      },
      {
        contributionCount: 8,
      },
      {
        contributionCount: 0,
      },
      {
        contributionCount: 0,
      },
      {
        contributionCount: 14,
      },
      {
        contributionCount: 6,
      },
    ],
  },
  {
    contributionDays: [
      {
        contributionCount: 13,
      },
      {
        contributionCount: 2,
      },
      {
        contributionCount: 10,
      },
      {
        contributionCount: 2,
      },
      {
        contributionCount: 1,
      },
      {
        contributionCount: 5,
      },
      {
        contributionCount: 13,
      },
    ],
  },
  {
    contributionDays: [
      {
        contributionCount: 1,
      },
      {
        contributionCount: 3,
      },
      {
        contributionCount: 6,
      },
      {
        contributionCount: 7,
      },
      {
        contributionCount: 2,
      },
      {
        contributionCount: 1,
      },
      {
        contributionCount: 5,
      },
    ],
  },
  {
    contributionDays: [
      {
        contributionCount: 11,
      },
    ],
  },
];

export const themes = {
  dracula: {
    bgColor: '44475a',
    borderColor: 'ffffff',
    color: 'f8f8f2',
    lineColor: 'ff79c6',
    pointColor: 'bd93f9',
  },
  gruvbox: {
    bgColor: '504945',
    borderColor: 'ffffff',
    color: 'd4be98',
    lineColor: 'd8a657',
    pointColor: 'e78a4e',
  },
  github: {
    bgColor: '293036',
    borderColor: 'ffffff',
    color: 'ffffff',
    lineColor: '9ecbff',
    pointColor: 'f97583',
  },
  rogue: {
    bgColor: '172030',
    borderColor: 'ffffff',
    color: 'a3b09a',
    lineColor: 'b18bb1',
    pointColor: 'c6797e',
  },
  xcode: {
    bgColor: '202124',
    borderColor: 'ffffff',
    color: 'fcfcfa',
    lineColor: 'c4e3ff',
    pointColor: 'ff8070',
  },
  redical: {
    bgColor: '141321',
    borderColor: 'ffffff',
    color: 'a9fef7',
    lineColor: 'fe428e',
    pointColor: 'f8d847',
  },
  coral: {
    bgColor: '9a3838',
    borderColor: 'ffffff',
    color: 'f9fae9',
    lineColor: 'f4e23d',
    pointColor: 'f4e7e7',
  },
  reactDark: {
    bgColor: '0d1117',
    borderColor: 'ffffff',
    color: '5bcdec',
    lineColor: '5bcdec',
    pointColor: 'ffffff',
  },
  nord: {
    bgColor: '2e3440',
    borderColor: 'ffffff',
    color: '88c0d0',
    lineColor: '88c0d0',
    pointColor: 'ffffff',
  },
  lucent: {
    bgColor: 'cccccc',
    borderColor: '000000',
    color: '000000',
    lineColor: 'ffd369',
    pointColor: 'faf3e0',
  },
  default: {
    bgColor: 'ffcfe9',
    borderColor: 'ffffff',
    color: '9e4c98',
    lineColor: '9e4c98',
    pointColor: '403d3d',
  },
};
