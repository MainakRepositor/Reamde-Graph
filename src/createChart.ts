const co = require('co');
const chartist = require('../node_modules/node-chartist/lib/chartist');
const generateChart = require('../node_modules/node-chartist/lib/chart');
const generateLegend = require('../node_modules/node-chartist/lib/legend');
const is = require('is_js');
const R = require('ramda');
const Ru = require('@panosoft/ramda-utils');

/**
 * Generate Chart HTML
 *
 * @param {String} type
 *        bar, line, pie
 * @param {Object} options
 *        Chartist options + axis*.title
 * @param {Object} data
 *        Chartist data object
 *
 * @return {Promise{String}} svg
 */

export const createGraph = R.curryN(
  3,
  co.wrap(function* (type: string, options: any, data: any) {
    const environment = yield chartist.initialize();
    const window = environment.window;
    const Chartist = environment.Chartist;
    // process options
    options = is.function(options) ? options(Chartist) : options;
    if (is.not.json(options))
      throw new TypeError(
        'options must be an object or a function that returns an object.'
      );
    options = Ru.defaults({ legend: false }, options);
    // create chart
    const chart = yield generateChart(Chartist, window, type, options, data);
    const legend = options.legend ? generateLegend(data) : '';
    return `${chart}${legend}`;
  })
);

export default { chartist };
