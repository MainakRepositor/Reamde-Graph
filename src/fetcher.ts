import axios from 'axios';
import {
  query,
  userDetails,
  week,
  contributionCount,
} from '../interfaces/interface';
import { fetchContribution, fetcher, gqlQuery } from '../types/types';

require('dotenv').config('../');

//GraphQl query to get everyday contributions as a response
export const graphqlQuery: gqlQuery = (username: string) => {
  return {
    query: `
      query userInfo($LOGIN: String!) {
       user(login: $LOGIN) {
         name
         contributionsCollection {
           contributionCalendar {
              totalContributions
              weeks {
                contributionDays {
                  contributionCount
                }
              }
            }
          }
        }
      },
    `,
    variables: {
      LOGIN: username,
    },
  };
};

const token: string | undefined = process.env.TOKEN;

const headers = {
  Authorization: `bearer ${token}`,
};

//Fetching data from GitHub GraphQl API
export const fetch: fetcher = (data: query) =>
  axios({
    url: 'https://api.github.com/graphql',
    method: 'POST',
    headers,
    data,
  });

export const fetchContributions: fetchContribution = async (
  username: string,
  graphqlQuery: gqlQuery, //Dependency Injection
  fetch: fetcher
) => {
  try {
    const apiResponse = await fetch(graphqlQuery(username));
    if (apiResponse.data.data.user === null)
      return `Can't fetch any contribution. Please check your username ðŸ˜¬`;
    else {
      let userData: userDetails = {
        contributions: [],
        name: apiResponse.data.data.user.name,
      };
      //filtering the week data from API response
      const weeks: week[] =
        apiResponse.data.data.user.contributionsCollection.contributionCalendar
          .weeks;
      //slicing last 6 weeks
      weeks.slice(weeks.length - 6, weeks.length).map((week: week) =>
        week.contributionDays.map((contributionCount: contributionCount) => {
          userData.contributions.push(contributionCount.contributionCount);
        })
      );

      //returning data of last 31 days
      const presentDay = new Date().getDay();
      userData.contributions = userData.contributions.slice(
        5 + presentDay,
        36 + presentDay
      );
      return userData;
    }
  } catch (error) {
    return error;
  }
};
