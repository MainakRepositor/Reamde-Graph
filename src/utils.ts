import { Request, Response } from 'express';
import { Card } from './GraphCards';
import { invalidUserSvg } from './svgs';
import {
  queryOption,
  colors,
  ParsedQs,
  userDetails,
} from '../interfaces/interface';
import { fetcher, gqlQuery } from '../types/types';
import { fetchContributions } from './fetcher';
import { selectColors } from '../styles/themes';

export const queryOptions = (queryString: ParsedQs): queryOption => {
  let area: boolean = false;
  let colors: colors;
  let theme: string = queryString.theme || 'default';

  if (String(queryString.area) === 'true') {
    area = true;
  }

  // Custom options for user
  colors = {
    bgColor: queryString.bg_color
      ? queryString.bg_color
      : selectColors(theme).bgColor,
    borderColor:
      String(queryString.hide_border) === 'true'
        ? '0000' // transparent
        : selectColors(theme).borderColor,
    color: queryString.color ? queryString.color : selectColors(theme).color,
    lineColor: queryString.line
      ? queryString.line
      : selectColors(theme).lineColor,
    pointColor: queryString.point
      ? queryString.point
      : selectColors(theme).pointColor,
  };

  const options: queryOption = {
    username: String(queryString.username),
    colors: colors,
    area: area,
  };

  return options;
};

const setHttpHeader = (res: Response, directivesAndAge: string): void => {
  res.setHeader('Cache-Control', `${directivesAndAge}`);
  res.set('Content-Type', 'image/svg+xml');
};

//HOF
export const getGraph = (graphqlQuery: gqlQuery, fetch: fetcher) => async (
  req: Request,
  res: Response
) => {
  try {
    const options: queryOption = queryOptions(req.query);

    const fetchCalendarData: userDetails | string = await fetchContributions(
      `${options.username}`,
      graphqlQuery,
      fetch
    );

    if (typeof fetchCalendarData === 'object') {
      const graph: Card = new Card(
        420,
        1200,
        options.colors,
        `${fetchCalendarData.name}'s Contribution Graph`,
        options.area
      );

      const getChart: string = await graph.chart(
        fetchCalendarData.contributions
      );

      setHttpHeader(res, 'public, max-age=1800');
      res.status(200).send(getChart);
    } else {
      setHttpHeader(res, 'no-store, max-age=0');
      res.send(invalidUserSvg(fetchCalendarData));
    }
  } catch (error) {
    setHttpHeader(res, 'no-store, max-age=0');
    res.send(invalidUserSvg('Something unexpected happened ðŸ’¥'));
  }
};

/* DO NOT CHANGE THE CODE BELOW */
export const getData = (graphqlQuery: gqlQuery, fetch: fetcher) => async (
  req: Request,
  res: Response
) => {
  try {
    const options: queryOption = queryOptions(req.query);

    const fetchCalendarData: userDetails | string = await fetchContributions(
      `${options.username}`,
      graphqlQuery,
      fetch
    );

    if (typeof fetchCalendarData === 'object') {
      res.status(200).send(fetchCalendarData);
    } else {
      res.send(invalidUserSvg(fetchCalendarData));
    }
  } catch (error) {
    res.send(invalidUserSvg('Something unexpected happened ðŸ’¥'));
  }
};
