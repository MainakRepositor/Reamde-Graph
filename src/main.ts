import express, { Application, Request, Response } from 'express';
import { getData, getGraph } from './utils';
import { graphqlQuery, fetch } from './fetcher';
import cors from 'cors';

const app: Application = express();
let port: string | number = process.env.PORT || 5100;

app.use(express.urlencoded({ extended: false }));
app.use(cors());

app.get('/', (_req: Request, res: Response) => {
  res.send(`<h1>GitHub Readme Activity Graph ðŸ“ˆ</h1>`);
});

//Get Graph
app.get('/graph', getGraph(graphqlQuery, fetch));
app.get('/data', getData(graphqlQuery, fetch));

app.listen(port, (): void => {
  console.log(`Server is Running in port ${port}`);
});
