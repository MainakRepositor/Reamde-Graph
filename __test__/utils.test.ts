import express from 'express';
import request from 'supertest';
import { queryOptions, getGraph } from '../src/utils';
import { createGraph } from '../src/createChart';
import {
  mockQueryCorrect,
  mockQueryIncorrect,
  mockFetchCorrect,
  mockFetchIncorrect,
} from './mockFunctions';
import { fakeQueryString, fakeQueryStringRes, options } from './fakeInputs';

//- Query Option Function (utils.ts) ✔
it('Query Options', () => {
  expect(fakeQueryString.map((arg) => queryOptions(arg))).toEqual(
    fakeQueryStringRes
  );
});

//Testing express routes
const fakeSurver = () => {
  const app = express();
  app.use(express.urlencoded({ extended: false }));
  return app;
};

describe('GET /graph with correct credential', () => {
  it('responding', (done) => {
    const app = fakeSurver();
    app.get('/graph', getGraph(mockQueryCorrect, mockFetchCorrect));
    request(app)
      .get('/graph?username=ashutosh00710')
      .expect('Content-Type', 'image/svg+xml; charset=utf-8')
      .expect('Cache-Control', 'public, max-age=1800')
      .expect(200, done);
  });
});

describe('GET /graph with incorrect credential', () => {
  it('responding', (done) => {
    const app = fakeSurver();
    app.get('/graph', getGraph(mockQueryIncorrect, mockFetchIncorrect));
    request(app)
      .get('/graph?username=')
      .expect('Content-Type', 'image/svg+xml; charset=utf-8')
      .expect('Cache-Control', 'no-store, max-age=0')
      .expect(200, done);
  });
});

//- Chart Function ([Promise] Inside Graph Cards Class) ✔
it('Graph Generation', async () => {
  expect.assertions(1);
  let contributions = [1, 2, 3];
  const graph = await createGraph('line', options, {
    labels: Array.from(Array(contributions.length).keys(), (day) => day + 1),
    series: [{ value: contributions }],
  });
  expect(graph).toMatchSnapshot();
});
