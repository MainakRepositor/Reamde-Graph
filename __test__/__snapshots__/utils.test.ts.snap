// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph Generation 1`] = `"<svg xmlns:ct=\\"http://gionkunz.github.com/chartist-js/ct\\" width=\\"10\\" height=\\"5\\" class=\\"ct-chart-line\\"><g class=\\"ct-grids\\"><line x1=\\"90\\" x2=\\"90\\" y1=\\"80\\" y2=\\"81\\" class=\\"ct-grid ct-horizontal\\"></line><line x1=\\"90.5\\" x2=\\"90.5\\" y1=\\"80\\" y2=\\"81\\" class=\\"ct-grid ct-horizontal\\"></line><line x1=\\"91\\" x2=\\"91\\" y1=\\"80\\" y2=\\"81\\" class=\\"ct-grid ct-horizontal\\"></line><line y1=\\"81\\" y2=\\"81\\" x1=\\"90\\" x2=\\"91\\" class=\\"ct-grid ct-vertical\\"></line></g><g><g class=\\"ct-series ct-series-a\\"><path d=\\"M90,81C90.167,80.833,90.333,80.667,90.5,80.5C90.667,80.333,90.833,80.167,91,80\\" class=\\"ct-line\\"></path><line x1=\\"90\\" y1=\\"81\\" x2=\\"90.01\\" y2=\\"81\\" class=\\"ct-point\\" ct:value=\\"1\\"></line><line x1=\\"90.5\\" y1=\\"80.5\\" x2=\\"90.51\\" y2=\\"80.5\\" class=\\"ct-point\\" ct:value=\\"2\\"></line><line x1=\\"91\\" y1=\\"80\\" x2=\\"91.01\\" y2=\\"80\\" class=\\"ct-point\\" ct:value=\\"3\\"></line></g></g><g class=\\"ct-labels\\"><text x=\\"85.5\\" y=\\"101\\" width=\\"0.5\\" height=\\"40\\" class=\\"ct-label ct-horizontal ct-end\\">1</text><text x=\\"86\\" y=\\"101\\" width=\\"0.5\\" height=\\"40\\" class=\\"ct-label ct-horizontal ct-end\\">2</text><text x=\\"86.5\\" y=\\"101\\" width=\\"30\\" height=\\"40\\" class=\\"ct-label ct-horizontal ct-end\\">3</text><text y=\\"85.5\\" x=\\"80\\" height=\\"30\\" width=\\"60\\" class=\\"ct-label ct-vertical ct-start\\">1</text></g><text class=\\"ct-axis-title ct-label\\" x=\\"90.5\\" y=\\"131\\" dominant-baseline=\\"text-after-edge\\" text-anchor=\\"middle\\">Days</text><text class=\\"ct-axis-title ct-label\\" x=\\"20\\" y=\\"80.5\\" transform=\\"rotate(-90, 20, 80.5)\\" dominant-baseline=\\"hanging\\" text-anchor=\\"middle\\">Contributions</text></svg>"`;
